Add parameter validation for ECP module

hanno-arm opened the pull request:
__Summary:__ Based on #1643, this PR adds parameter validation for the ECP module.

mpg commented on include/mbedtls/ecp.h:
```diff
@@ -597,15 +615,21 @@ int mbedtls_ecp_point_read_string( mbedtls_ecp_point *P, int radix,
+ * \param format    The point format. This must be either
+ *                  #MBEDTLS_ECP_PF_COMPRESSED or #MBEDTLS_ECP_PF_UNCOMPRESSED.
+ * \param olen      The address at which to store the length of
+ *                  the output in Bytes.
```
Minor: we might add "This must not be NULL", or "This must point to a valid size_t object". Otherwise people might think that passing NULL is OK if they're not interested in the exact length or the output.

hanno-arm replied:
Fixed.

mpg commented on include/mbedtls/ecp.h:
```diff
@@ -649,99 +679,122 @@ int mbedtls_ecp_point_read_binary( const mbedtls_ecp_group *grp, mbedtls_ecp_poi
 /**
- * \brief           This function performs multiplication of a point by
- *                  an integer: \p R = \p m * \p P.
+ * \brief           This function performs a scalaar multiplication of a point
```
Typo: scalaar

hanno-arm replied:
Fixed.

mpg commented on tests/suites/test_suite_ecp.function:
```diff
@@ -13,6 +13,315 @@
+    TEST_INVALID_PARAM( mbedtls_ecp_keypair_init( NULL ) );
+    TEST_INVALID_PARAM( mbedtls_ecp_point_init( NULL ) );
+
+    TEST_VALID_PARAM( mbedtls_ecp_point_free( NULL ) );
```
Note: I think these should ideally go into a separate testing function that doesn't depend on `MBEDTLS_CHECK_PARAMS` but I don't think that's a blocker.

hanno-arm replied:
Fixed.

hanno-arm commented:
Thanks @mpg for your review. I believe I have addressed your comments. Could you please re-review?

hanno-arm commented:
@mpg I rebased this on top of the current #1643 and removed a few 'no-op's which @sbutcher-arm requested we shouldn't use.

hanno-arm commented:
@mpg Please ignore the first fauly rebase, and see https://github.com/ARMmbed/mbedtls/compare/b993ee4d65372b37cd26c68e487c23648724f239..486f1b33d72f960f7cb8b9fbd36d992a26697692 for the diff between the state before the two forced pushes and now.

gilles-peskine-arm commented on library/ecp.c:
```diff
@@ -2333,6 +2383,10 @@ int mbedtls_ecp_mul( mbedtls_ecp_group *grp, mbedtls_ecp_point *R,
              const mbedtls_mpi *m, const mbedtls_ecp_point *P,
              int (*f_rng)(void *, unsigned char *, size_t), void *p_rng )
 {
+    ECP_VALIDATE_RET( grp != NULL );
```
Not a blocker: `mbedtls_ecp_mul_restartable` will do the exact same checks, so the ones here are redundant. Also applies to other functions that are trivial wrappers around some other function (`xxx` → `xxx_restartable`, `mbedtls_ecp_gen_keypair` → `mbedtls_ecp_gen_keypair_base` except for `grp`, `mbedtls_ecp_gen_keypair_base` → `mbedtls_ecp_gen_privkey`).

gilles-peskine-arm commented on tests/suites/test_suite_ecp.function:
```diff
@@ -13,6 +13,326 @@
+
+    TEST_INVALID_PARAM( mbedtls_ecp_point_init( NULL ) );
+    TEST_INVALID_PARAM( mbedtls_ecp_keypair_init( NULL ) );
+    TEST_INVALID_PARAM( mbedtls_ecp_point_init( NULL ) );
```
Duplicate of two lines above

hanno-arm replied:
This should have been `mbedtls_ecp_group_init()`.

hanno-arm commented:
> Missing null check in mbedtls_ecp_restart_init,

Fixed.

>  Missing null check in mbedtls_ecp_check_budget.

The function only needs to check the group context, which it does.

> Missing null check in mbedtls_ecp_curve_info_from_name.

The current parameter validation macros don't support functions returning pointers. I added an unconditional check. This might be preferable in any case, since `strcmp()` is undefined for NULL inputs.

> Missing test cases for mbedtls_ecp_check_budget, mbedtls_ecp_group_init, mbedtls_gen_privkey, mbedtls_ecp_check_pub_priv

Fixed.

> plus mbedtls_ecp_curve_info_from_name

No test added because the added check is unconditional and doesn't use parameter validation.

mpg commented:
@hanno-arm Did you forget to push after the latest fixes? I can see you replied to Gilles that some things were fixed, but I can't see any new commits since his review, and the first issue does not appear to be fixed in the tip of the branch as I see it.

hanno-arm commented:
@mpg Yes, sorry for that, and thanks for noticing.

mpg commented:
CI only failed in the known flaky timing test, so as good as a pass.

sbutcher-arm commented:
Reviewed, approved, passes CI, but needs rebasing.

Removing the 'needs: preceding PR' label, as PR #1643 has now been merged.

mpg commented:
@sbutcher-arm I don't think this branch needs rebasing, it has no conflicts and was based on a recent enough version to pass CI.

sbutcher-arm commented:
@mpg - Yes, you're right. Thanks for changing the base - it has removed any commits from PR #1643.

Jenkins CI fails only with the BSD timing test, which is a known problem. The PR has been reviewed and approved and passed CI, so labelling as 'ready to merge'.
